// spring-framework/dist/test/packagePoc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.TestBean_Drawable = exports.TestBean_Business = exports.TestBean_Core = exports.TestContext_Drawable = exports.TestContext_Business = exports.TestContext_Core = exports.TestBean_Drawable_Token = exports.TestBean_Business_Token = exports.TestBean_Core_Token = exports.packagePoc = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _orgSpringframework = require("../org.springframework.beans");

var _orgSpringframework2 = require("../org.springframework.context");

var __extends = undefined && undefined.__extends || function () {
    var _extendStatics = function extendStatics(d, b) {
        _extendStatics = Object.setPrototypeOf || { __proto__: [] } instanceof Array && function (d, b) {
            d.__proto__ = b;
        } || function (d, b) {
            for (var p in b) {
                if (b.hasOwnProperty(p)) d[p] = b[p];
            }
        };
        return _extendStatics(d, b);
    };
    return function (d, b) {
        _extendStatics(d, b);
        function __() {
            this.constructor = d;
        }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
}();
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
    var c = arguments.length,
        r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
        d;
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) {
        if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    }return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = undefined && undefined.__metadata || function (k, v) {
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};

//
// Tokens
//
var packagePoc = exports.packagePoc = "com.test.poc";
var TestBean_Core_Token = exports.TestBean_Core_Token = Symbol("TestBean3_Core");
var TestBean_Business_Token = exports.TestBean_Business_Token = Symbol("TestBean3_Business");
var TestBean_Drawable_Token = exports.TestBean_Drawable_Token = Symbol("TestBean3_Drawable");
//
// Contexts
//
var TestContext_Core = /** @class */function (_super) {
    __extends(TestContext_Core, _super);
    function TestContext_Core() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestContext_Core.prototype.configure = function () {
        _super.prototype.configure.call(this, new Map([]));
    };
    return TestContext_Core;
}(_orgSpringframework2.AbstractApplicationContext);
exports.TestContext_Core = TestContext_Core;

var TestContext_Business = /** @class */function (_super) {
    __extends(TestContext_Business, _super);
    function TestContext_Business() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestContext_Business.prototype.configure = function () {
        _super.prototype.configure.call(this, new Map([[TestBean_Core_Token, TestBean_Business_Token]]));
    };
    return TestContext_Business;
}(_orgSpringframework2.AbstractApplicationContext);
exports.TestContext_Business = TestContext_Business;

var TestContext_Drawable = /** @class */function (_super) {
    __extends(TestContext_Drawable, _super);
    function TestContext_Drawable() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestContext_Drawable.prototype.configure = function () {
        _super.prototype.configure.call(this, new Map([]));
    };
    return TestContext_Drawable;
}(_orgSpringframework2.AbstractApplicationContext);
exports.TestContext_Drawable = TestContext_Drawable;
//
// Beans
//

var TestBean_Core = /** @class */function () {
    function TestBean_Core() {}
    TestBean_Core.prototype.getName = function () {
        return "Core";
    };
    TestBean_Core = __decorate([(0, _orgSpringframework2.Bean)(TestBean_Core_Token, { scope: "singleton" })], TestBean_Core);
    return TestBean_Core;
}();
exports.TestBean_Core = TestBean_Core;

var TestBean_Business = /** @class */function (_super) {
    __extends(TestBean_Business, _super);
    function TestBean_Business() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestBean_Business.prototype.getName = function () {
        return "Business";
    };
    TestBean_Business = __decorate([(0, _orgSpringframework2.Bean)(TestBean_Business_Token, { scope: "singleton" })], TestBean_Business);
    return TestBean_Business;
}(TestBean_Core);
exports.TestBean_Business = TestBean_Business;

var TestBean_Drawable = /** @class */function () {
    function TestBean_Drawable() {}
    TestBean_Drawable.prototype.getName = function () {
        return "Drawable";
    };
    __decorate([(0, _orgSpringframework.Autowired)(TestBean_Core_Token), __metadata("design:type", TestBean_Core)], TestBean_Drawable.prototype, "core", void 0);
    TestBean_Drawable = __decorate([(0, _orgSpringframework2.Bean)(TestBean_Drawable_Token, { scope: "prototype" })], TestBean_Drawable);
    return TestBean_Drawable;
}();
exports.TestBean_Drawable = TestBean_Drawable;
