// spring-framework/dist/test/package1.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.TestBean1_4_Impl = exports.TestBean1_4 = exports.TestBean1_3 = exports.TestBean1_2 = exports.TestBean1_1 = exports.TestContext1_2 = exports.TestContext1_1 = exports.TestBean1_4_Token_Impl = exports.TestBean1_4_Token = exports.TestBean1_3_Token = exports.TestBean1_2_Token = exports.TestBean1_1_Token = exports.package1 = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _orgSpringframework = require("../org.springframework.beans");

var _orgSpringframework2 = require("../org.springframework.context");

var __extends = undefined && undefined.__extends || function () {
    var _extendStatics = function extendStatics(d, b) {
        _extendStatics = Object.setPrototypeOf || { __proto__: [] } instanceof Array && function (d, b) {
            d.__proto__ = b;
        } || function (d, b) {
            for (var p in b) {
                if (b.hasOwnProperty(p)) d[p] = b[p];
            }
        };
        return _extendStatics(d, b);
    };
    return function (d, b) {
        _extendStatics(d, b);
        function __() {
            this.constructor = d;
        }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
}();
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
    var c = arguments.length,
        r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
        d;
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) {
        if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    }return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = undefined && undefined.__metadata || function (k, v) {
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};

//
// Tokens
//
var package1 = exports.package1 = "org.springframework.test.1";
var TestBean1_1_Token = exports.TestBean1_1_Token = Symbol("TestBean1_1");
var TestBean1_2_Token = exports.TestBean1_2_Token = Symbol("TestBean1_2");
var TestBean1_3_Token = exports.TestBean1_3_Token = Symbol("TestBean1_3");
var TestBean1_4_Token = exports.TestBean1_4_Token = Symbol("TestBean1_4");
var TestBean1_4_Token_Impl = exports.TestBean1_4_Token_Impl = Symbol("TestBean1_4_Token_Impl");
//
// Contexts
//
var TestContext1_1 = /** @class */function (_super) {
    __extends(TestContext1_1, _super);
    function TestContext1_1() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestContext1_1.prototype.configure = function () {
        _super.prototype.configure.call(this, new Map([[package1 + ".1_3", TestBean1_3_Token], [package1 + ".1_3-proxy", package1 + ".1_3"], [TestBean1_4_Token, TestBean1_4_Token_Impl]]));
    };
    return TestContext1_1;
}(_orgSpringframework2.AbstractApplicationContext);
exports.TestContext1_1 = TestContext1_1;

var TestContext1_2 = /** @class */function (_super) {
    __extends(TestContext1_2, _super);
    function TestContext1_2() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return TestContext1_2;
}(TestContext1_1);
exports.TestContext1_2 = TestContext1_2;
//
// Beans
//

var TestBean1_1 = /** @class */function () {
    function TestBean1_1() {}
    TestBean1_1.prototype.getName = function () {
        return "Bean 1_1";
    };
    TestBean1_1 = __decorate([(0, _orgSpringframework2.Bean)(TestBean1_1_Token)], TestBean1_1);
    return TestBean1_1;
}();
exports.TestBean1_1 = TestBean1_1;

var TestBean1_2 = /** @class */function () {
    function TestBean1_2() {}
    // @Resource("https://...", { required: false }) public resource1: Promise<any>;
    TestBean1_2.prototype.getName = function () {
        return "Bean 1_2";
    };
    __decorate([(0, _orgSpringframework.Autowired)(TestBean1_1_Token), __metadata("design:type", TestBean1_1)], TestBean1_2.prototype, "bean1", void 0);
    TestBean1_2 = __decorate([(0, _orgSpringframework2.Bean)(TestBean1_2_Token)], TestBean1_2);
    return TestBean1_2;
}();
exports.TestBean1_2 = TestBean1_2;

var TestBean1_3 = /** @class */function () {
    function TestBean1_3() {}
    TestBean1_3.prototype.getName = function () {
        return "Bean 1_3";
    };
    __decorate([(0, _orgSpringframework.Autowired)(TestBean1_1_Token), __metadata("design:type", TestBean1_1)], TestBean1_3.prototype, "bean1", void 0);
    __decorate([(0, _orgSpringframework.Autowired)(TestBean1_2_Token), __metadata("design:type", TestBean1_2)], TestBean1_3.prototype, "bean2", void 0);
    TestBean1_3 = __decorate([(0, _orgSpringframework2.Bean)(TestBean1_3_Token)], TestBean1_3);
    return TestBean1_3;
}();
exports.TestBean1_3 = TestBean1_3;

var TestBean1_4 = /** @class */function () {
    function TestBean1_4() {}
    TestBean1_4.prototype.getName = function () {
        return "Bean 1_4";
    };
    __decorate([(0, _orgSpringframework.Autowired)(TestBean1_1_Token), __metadata("design:type", TestBean1_1)], TestBean1_4.prototype, "bean1", void 0);
    TestBean1_4 = __decorate([(0, _orgSpringframework2.Bean)(TestBean1_4_Token)], TestBean1_4);
    return TestBean1_4;
}();
exports.TestBean1_4 = TestBean1_4;

var TestBean1_4_Impl = /** @class */function (_super) {
    __extends(TestBean1_4_Impl, _super);
    function TestBean1_4_Impl() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestBean1_4_Impl.prototype.getName = function () {
        return "Bean 1_4_Impl";
    };
    TestBean1_4_Impl.prototype.getFullName = function () {
        return this.getName() + " " + this.bean1.getName() + " " + this.bean2.getName();
    };
    __decorate([(0, _orgSpringframework.Autowired)(TestBean1_2_Token), __metadata("design:type", TestBean1_2)], TestBean1_4_Impl.prototype, "bean2", void 0);
    TestBean1_4_Impl = __decorate([(0, _orgSpringframework2.Bean)(TestBean1_4_Token_Impl)], TestBean1_4_Impl);
    return TestBean1_4_Impl;
}(TestBean1_4);
exports.TestBean1_4_Impl = TestBean1_4_Impl;
