// spring-framework/dist/test/package2.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.TestBean2_4 = exports.TestBean2_3 = exports.TestBean2_2 = exports.TestBean2_1 = exports.TestContext2_2 = exports.TestContext2_1 = exports.TestBean2_Unknown_Token = exports.TestBean2_Value_Token = exports.TestBean2_1_TokenProxy = exports.TestBean2_4_Token = exports.TestBean2_3_Token = exports.TestBean2_2_Token = exports.TestBean2_1_Token = exports.package2 = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _javax = require("../javax");

var _orgSpringframework = require("../org.springframework.beans");

var _orgSpringframework2 = require("../org.springframework.context");

var __extends = undefined && undefined.__extends || function () {
    var _extendStatics = function extendStatics(d, b) {
        _extendStatics = Object.setPrototypeOf || { __proto__: [] } instanceof Array && function (d, b) {
            d.__proto__ = b;
        } || function (d, b) {
            for (var p in b) {
                if (b.hasOwnProperty(p)) d[p] = b[p];
            }
        };
        return _extendStatics(d, b);
    };
    return function (d, b) {
        _extendStatics(d, b);
        function __() {
            this.constructor = d;
        }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
}();
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
    var c = arguments.length,
        r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
        d;
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) {
        if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    }return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = undefined && undefined.__metadata || function (k, v) {
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};

//
// Tokens
//
var package2 = exports.package2 = "org.springframework.test.2";
var TestBean2_1_Token = exports.TestBean2_1_Token = Symbol("TestBean2_1");
var TestBean2_2_Token = exports.TestBean2_2_Token = Symbol("TestBean2_2");
var TestBean2_3_Token = exports.TestBean2_3_Token = Symbol("TestBean2_3");
var TestBean2_4_Token = exports.TestBean2_4_Token = Symbol("TestBean2_4");
var TestBean2_1_TokenProxy = exports.TestBean2_1_TokenProxy = Symbol("TestBean2_1_Proxy");
var TestBean2_Value_Token = exports.TestBean2_Value_Token = Symbol("TestBean2_Value");
var TestBean2_Unknown_Token = exports.TestBean2_Unknown_Token = Symbol("TestBean2_Unknown");
//
// Contexts
//
var TestContext2_1 = /** @class */function (_super) {
    __extends(TestContext2_1, _super);
    function TestContext2_1() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestContext2_1.prototype.configure = function () {
        var symbol = Symbol();
        _super.prototype.configure.call(this, new Map([[TestBean2_1_TokenProxy, TestBean2_1_Token], [package2 + ".2_1", TestBean2_1_Token], [package2 + ".2_1-proxy", package2 + ".2_1"], [package2 + ".factory-prototype", _orgSpringframework.FactoryBean.of(function () {
            return Symbol();
        })], [package2 + ".factory-singleton", _orgSpringframework.FactoryBean.of(function () {
            return symbol;
        })]]));
    };
    return TestContext2_1;
}(_orgSpringframework2.AbstractApplicationContext);
exports.TestContext2_1 = TestContext2_1;

var TestContext2_2 = /** @class */function (_super) {
    __extends(TestContext2_2, _super);
    function TestContext2_2() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestContext2_2.prototype.configure = function () {
        _super.prototype.configure.call(this, new Map([[package2 + ".redirect", TestBean2_1_Token]]));
    };
    return TestContext2_2;
}(_orgSpringframework2.AbstractApplicationContext);
exports.TestContext2_2 = TestContext2_2;
//
// Beans
//

var TestBean2_1 = /** @class */function () {
    function TestBean2_1() {
        this.initHookCalled = false;
        this.deinitHookCalled = false;
    }
    TestBean2_1.prototype.getName = function () {
        return "Bean 2_1";
    };
    TestBean2_1.prototype.construct = function () {
        this.initHookCalled = true;
    };
    TestBean2_1.prototype.destruct = function () {
        this.deinitHookCalled = true;
    };
    __decorate([_javax.PostConstruct, __metadata("design:type", Function), __metadata("design:paramtypes", []), __metadata("design:returntype", void 0)], TestBean2_1.prototype, "construct", null);
    __decorate([_javax.PreDestroy, __metadata("design:type", Function), __metadata("design:paramtypes", []), __metadata("design:returntype", void 0)], TestBean2_1.prototype, "destruct", null);
    TestBean2_1 = __decorate([(0, _orgSpringframework2.Bean)(TestBean2_1_Token, { scope: "singleton" })], TestBean2_1);
    return TestBean2_1;
}();
exports.TestBean2_1 = TestBean2_1;

var TestBean2_2 = /** @class */function () {
    function TestBean2_2() {
        this.initHookCalled = false;
        this.deinitHookCalled = false;
    }
    TestBean2_2.prototype.getName = function () {
        return "Global bean 2_2";
    };
    TestBean2_2.prototype.construct = function () {
        this.initHookCalled = true;
    };
    TestBean2_2.prototype.destruct = function () {
        this.deinitHookCalled = true;
    };
    __decorate([_javax.PostConstruct, __metadata("design:type", Function), __metadata("design:paramtypes", []), __metadata("design:returntype", void 0)], TestBean2_2.prototype, "construct", null);
    __decorate([_javax.PreDestroy, __metadata("design:type", Function), __metadata("design:paramtypes", []), __metadata("design:returntype", void 0)], TestBean2_2.prototype, "destruct", null);
    TestBean2_2 = __decorate([(0, _orgSpringframework2.Bean)(TestBean2_2_Token, { scope: "global" })], TestBean2_2);
    return TestBean2_2;
}();
exports.TestBean2_2 = TestBean2_2;

var TestBean2_3 = /** @class */function () {
    function TestBean2_3() {
        this.initHookCalled = false;
        this.deinitHookCalled = false;
        this.constructorCalled = false;
        this.constructorCalled = true;
    }
    TestBean2_3.prototype.getName = function () {
        return "Prototype bean 2_3";
    };
    TestBean2_3.prototype.construct = function () {
        this.initHookCalled = true;
    };
    TestBean2_3.prototype.destruct = function () {
        this.deinitHookCalled = true;
    };
    TestBean2_3 = __decorate([(0, _orgSpringframework2.Bean)(TestBean2_3_Token, { scope: "prototype" }), __metadata("design:paramtypes", [])], TestBean2_3);
    return TestBean2_3;
}();
exports.TestBean2_3 = TestBean2_3;

var TestBean2_4 = /** @class */function (_super) {
    __extends(TestBean2_4, _super);
    function TestBean2_4() {
        var _this = _super.call(this) || this;
        _this.initHookCalledChild = false;
        _this.deinitHookCalledChild = false;
        _this.constructorCalledChild = false;
        _this.constructorCalledChild = true;
        return _this;
    }
    TestBean2_4.prototype.getName = function () {
        return "Prototype bean 2_4";
    };
    TestBean2_4.prototype.construct = function () {
        _super.prototype.construct.call(this);
        this.initHookCalledChild = true;
    };
    TestBean2_4.prototype.destruct = function () {
        _super.prototype.destruct.call(this);
        this.deinitHookCalledChild = true;
    };
    __decorate([(0, _orgSpringframework.Autowired)(TestBean2_Value_Token, { required: false }), __metadata("design:type", Object)], TestBean2_4.prototype, "valueBean", void 0);
    __decorate([(0, _orgSpringframework.Autowired)(TestBean2_Unknown_Token, { required: false }), __metadata("design:type", void 0)], TestBean2_4.prototype, "unknownBean", void 0);
    __decorate([_javax.PostConstruct, __metadata("design:type", Function), __metadata("design:paramtypes", []), __metadata("design:returntype", void 0)], TestBean2_4.prototype, "construct", null);
    __decorate([_javax.PreDestroy, __metadata("design:type", Function), __metadata("design:paramtypes", []), __metadata("design:returntype", void 0)], TestBean2_4.prototype, "destruct", null);
    TestBean2_4 = __decorate([(0, _orgSpringframework2.Bean)(TestBean2_4_Token, { scope: "prototype" }), __metadata("design:paramtypes", [])], TestBean2_4);
    return TestBean2_4;
}(TestBean2_3);
exports.TestBean2_4 = TestBean2_4;
