// spring-framework/dist/test/packageContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.TestBean_1 = exports.TestContext = exports.TestBean_1_Token = exports.packagePoc = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _orgSpringframework = require("../org.springframework.context");

var _rxjava = require("../rxjava");

var __extends = undefined && undefined.__extends || function () {
    var _extendStatics = function extendStatics(d, b) {
        _extendStatics = Object.setPrototypeOf || { __proto__: [] } instanceof Array && function (d, b) {
            d.__proto__ = b;
        } || function (d, b) {
            for (var p in b) {
                if (b.hasOwnProperty(p)) d[p] = b[p];
            }
        };
        return _extendStatics(d, b);
    };
    return function (d, b) {
        _extendStatics(d, b);
        function __() {
            this.constructor = d;
        }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
}();
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
    var c = arguments.length,
        r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
        d;
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) {
        if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    }return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = undefined && undefined.__metadata || function (k, v) {
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};

//
// Tokens
//
var packagePoc = exports.packagePoc = "org.springframework.test.context";
var TestBean_1_Token = exports.TestBean_1_Token = Symbol("TestBean_1");
//
// Contexts
//
var TestContext = /** @class */function (_super) {
    __extends(TestContext, _super);
    function TestContext() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestContext.prototype.configure = function () {
        _super.prototype.configure.call(this, new Map([]));
    };
    return TestContext;
}(_orgSpringframework.AbstractApplicationContext);
exports.TestContext = TestContext;
//
// Beans
//

var TestBean_1 = /** @class */function () {
    function TestBean_1() {
        var _this = this;
        this.eventTriggered = false;
        this.disposer = function () {
            return _this.eventTriggered = null;
        };
    }
    TestBean_1.prototype.onApplicationEvent = function (event) {
        this.eventTriggered = true;
    };
    __decorate([_rxjava.Disposable, __metadata("design:type", Function)], TestBean_1.prototype, "disposer", void 0);
    __decorate([(0, _orgSpringframework.EventListener)(_orgSpringframework.ApplicationContextEvent), __metadata("design:type", Function), __metadata("design:paramtypes", [_orgSpringframework.ApplicationEvent]), __metadata("design:returntype", void 0)], TestBean_1.prototype, "onApplicationEvent", null);
    TestBean_1 = __decorate([(0, _orgSpringframework.Bean)(TestBean_1_Token, { scope: "singleton" })], TestBean_1);
    return TestBean_1;
}();
exports.TestBean_1 = TestBean_1;
