// spring-framework/dist/test/packageContextLoader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.applicationContextJsonFile = exports.Bean_1 = exports.JsonContext = exports.Bean_2_Token = exports.Bean_1_Token = exports.packagePoc = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _orgSpringframework = require("../org.springframework.context");

var _jsonUrlApplicationContext = require("../org.springframework.context/jsonUrlApplicationContext");

var _javax = require("../javax");

var __extends = undefined && undefined.__extends || function () {
    var _extendStatics = function extendStatics(d, b) {
        _extendStatics = Object.setPrototypeOf || { __proto__: [] } instanceof Array && function (d, b) {
            d.__proto__ = b;
        } || function (d, b) {
            for (var p in b) {
                if (b.hasOwnProperty(p)) d[p] = b[p];
            }
        };
        return _extendStatics(d, b);
    };
    return function (d, b) {
        _extendStatics(d, b);
        function __() {
            this.constructor = d;
        }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
}();
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
    var c = arguments.length,
        r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
        d;
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) {
        if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    }return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = undefined && undefined.__metadata || function (k, v) {
    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};

//
// Tokens
//
var packagePoc = exports.packagePoc = "org.springframework.test.contextLoader";
var Bean_1_Token = exports.Bean_1_Token = Symbol("Bean_1");
var Bean_2_Token = exports.Bean_2_Token = Symbol("Bean_2");
//
// Contexts
//
var JsonContext = /** @class */function (_super) {
    __extends(JsonContext, _super);
    function JsonContext() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    JsonContext.prototype.configure = function () {
        _super.prototype.configure.call(this, new Map([[packagePoc + ".bean1", Bean_1_Token], [packagePoc + ".bean2", Bean_2_Token]]));
        _super.prototype.configureBeansDefinitions.call(this);
    };
    JsonContext.prototype.start = function () {
        _super.prototype.startSchemaBeans.call(this);
    };
    return JsonContext;
}(_jsonUrlApplicationContext.JsonUrlApplicationContext);
exports.JsonContext = JsonContext;
//
// Beans
//

var Bean_1 = /** @class */function () {
    function Bean_1() {}
    __decorate([(0, _javax.Resource)("http://applicationContext.json"), __metadata("design:type", Object)], Bean_1.prototype, "schema", void 0);
    Bean_1 = __decorate([(0, _orgSpringframework.Bean)(Bean_1_Token)], Bean_1);
    return Bean_1;
}();
exports.Bean_1 = Bean_1;
//
// applicationContext.json
//

var applicationContextJsonFile = exports.applicationContextJsonFile = {
    beans: [{ id: packagePoc + ".bean1", optional: false }, { id: packagePoc + ".bean2", optional: true }],
    beansDefinition: [{ abstraction: packagePoc + ".name1", support: packagePoc + ".bean1" }, { abstraction: packagePoc + ".name2", support: packagePoc + ".bean2" }]
};
