// spring-framework/dist/javax/transaction.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.TransactionRequiredException = exports.TransactionException = exports.ETransactionStatus = undefined;

var _java = require("../java");

var __extends = undefined && undefined.__extends || function () {
    var _extendStatics = function extendStatics(d, b) {
        _extendStatics = Object.setPrototypeOf || { __proto__: [] } instanceof Array && function (d, b) {
            d.__proto__ = b;
        } || function (d, b) {
            for (var p in b) {
                if (b.hasOwnProperty(p)) d[p] = b[p];
            }
        };
        return _extendStatics(d, b);
    };
    return function (d, b) {
        _extendStatics(d, b);
        function __() {
            this.constructor = d;
        }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
}();
var ETransactionStatus = exports.ETransactionStatus = undefined;
(function (ETransactionStatus) {
    ETransactionStatus[ETransactionStatus["closed"] = 0] = "closed";
    ETransactionStatus[ETransactionStatus["opened"] = 1] = "opened";
    ETransactionStatus[ETransactionStatus["suspended"] = 2] = "suspended";
    ETransactionStatus[ETransactionStatus["commited"] = 3] = "commited";
    ETransactionStatus[ETransactionStatus["rollbacked"] = 4] = "rollbacked";
})(ETransactionStatus || (exports.ETransactionStatus = ETransactionStatus = {}));
var TransactionException = /** @class */function (_super) {
    __extends(TransactionException, _super);
    function TransactionException() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return TransactionException;
}(_java.Exception);
exports.TransactionException = TransactionException;

var TransactionRequiredException = /** @class */function (_super) {
    __extends(TransactionRequiredException, _super);
    function TransactionRequiredException() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return TransactionRequiredException;
}(_java.Exception);
exports.TransactionRequiredException = TransactionRequiredException;
