// spring-framework/dist/org.springframework.transaction/abstractTransactionManager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.AbstractTransactionManager = undefined;

var _transaction = require("../javax/transaction");

var AbstractTransactionManager = /** @class */function () {
    /* constructor */
    function AbstractTransactionManager() {
        this.semaphore = 0;
    }
    //
    // Api
    //
    AbstractTransactionManager.prototype.begin = function () {
        this.semaphore += 1;
    };
    AbstractTransactionManager.prototype.commit = function () {
        this.semaphore -= 1;
        if (!this.transaction) {
            throw new _transaction.TransactionRequiredException();
        }
        if (this.semaphore === 0) {
            this.transaction.commit();
        }
    };
    AbstractTransactionManager.prototype.rollback = function () {
        this.semaphore -= 1;
        if (!this.transaction) {
            throw new _transaction.TransactionRequiredException();
        }
        if (this.semaphore === 0) {
            this.transaction.rollback();
        }
    };
    AbstractTransactionManager.prototype.suspend = function () {
        this.suspended = true;
    };
    AbstractTransactionManager.prototype.resume = function () {
        this.suspended = false;
    };
    AbstractTransactionManager.prototype.getStatus = function () {
        return this.transaction ? this.transaction.getStatus() : _transaction.ETransactionStatus.closed;
    };
    AbstractTransactionManager.prototype.getTransaction = function () {
        return this.transaction;
    };
    return AbstractTransactionManager;
}();
exports.AbstractTransactionManager = AbstractTransactionManager;
