// spring-framework/dist/org.springframework.beans/annotations.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.dependenciesToken = undefined;
exports.Autowired = Autowired;

var _exception = require("../java/exception");

var __read = undefined && undefined.__read || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o),
        r,
        ar = [],
        e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) {
            ar.push(r.value);
        }
    } catch (error) {
        e = { error: error };
    } finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        } finally {
            if (e) throw e.error;
        }
    }
    return ar;
};
var __spread = undefined && undefined.__spread || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) {
        ar = ar.concat(__read(arguments[i]));
    }return ar;
};
var dependenciesToken = exports.dependenciesToken = Symbol();
//
// @Autowired
//
function Autowired(wishedBean, params) {
    params = params || {};
    params.required = params.required == null ? false : params.required;
    if (wishedBean == null) {
        throw new _exception.Exception("[@Autowired]: wished bean should not be null");
    }
    return function (target, propertyKey) {
        var destination = target.constructor;
        var dependencies = Reflect.getMetadata(dependenciesToken, destination) || [];
        Reflect.defineMetadata(dependenciesToken, __spread(dependencies, [{
            property: propertyKey,
            wishedBean: wishedBean,
            required: params.required,
            resolve: params.resolve
        }]), destination);
    };
}
